### ISSUE TITLE
ckEditor footnotes

### ISSUE THREAD

**Issue by loicjaouen** [dhlab-basel/Knora#528]
A project in Lausanne require footnotes in a textarea.
ckEditor is supposed to do that.

**Comment by loicjaouen** [comment]
following #526 except that I add footnotes in the addon lot.

**Comment by loicjaouen** [comment]
additionnal addons:
```
CKEditor Footnotes
Editor Resize
Horizontal Rule
Image
Image Paste
Image Resize
Paste From Excel
Paste from Word
Remove Format
Source Editing Area
```

before opening `samples/index.html` add to it the dependency to `jquery`:

```
	<script src="http://code.jquery.com/jquery-2.1.4.min.js"></script>
	<script src="../ckeditor.js"></script>
	<script src="js/sample.js"></script>
```

and here we are with a 4.7 ckeditor with footnotes.

**Comment by loicjaouen** [comment]
<img width="982" alt="capture d ecran 2017-06-01 a 17 28 58" src="https://cloud.githubusercontent.com/assets/3933654/26715225/2ad60284-4775-11e7-8dd7-61672989ec80.png">

**Comment by musicEnfanthen** [comment]
This is really great! Webern project, and I guess a lot more projects, also need footnotes.

How is this stored in the database? As Standoff like the other formatting options (italics, bold etc.)?

Do you see any possibility to add "Table" options to the editor, too?

**Comment by benjamingeer** [comment]
Footnotes are stored as standoff. @tobiasschweizer could you explain how you've mapped them for BEOL?

**Comment by loicjaouen** [comment]
it's great to feel momentum :)

> How is this stored in the database? As Standoff like the other formatting options (italics, bold etc.)?

Exactly.

On the front-end's side, ckEditor and its add-ons generates html code, that we might change a bit to make it fit knora's standoff xml parser (like in 49d5dc732e1d9152625e2ffcc999d0c47b81ace0).

On the back-end's side, knora's standoff xml parser splits it and use a defined [mapping](https://github.com/dhlab-basel/Knora/blob/develop/knora-ontologies/standoff-data.ttl) to match the xml/html tags to standoff semantic counter-part.

We can specify a different mapping alongside the xml code in the front-end when calling knora's standoff parser.

But we still have to discuss how we fix different ckEditor settings (ckEditor version + plugins).

> Do you see any possibility to add "Table" options to the editor, too?

Absolutely, the same way as for footnotes... hopefully this issue will reflect the process in extended details.

**Comment by loicjaouen** [comment]
Moving this version of ckEditor in salsah file hierarchy as `ckeditor_4.7.0-footnotes` without changing the config in `jquery.htmleditor.js` doesn't break anything.

Changing the tool bar; even if the footnotes button shows up in the sample as shown on the screenshot above, it doesn't show up in the config editor.

This is because the tool bar configuration tool doesn't recognize the section "others" that includes the footnotes plugin:

```json
config.toolbarGroups = [
		//...//
		{ name: 'tools' },
		{ name: 'others' },
		{ name: 'about' }
	];
```

Working out the config manually.

Including it into `jquery.htmleditor.js`.

There is no error in the console, but the result is a bit different than what `samples/index.html` shows:
<img width="434" alt="capture d ecran 2017-06-02 a 10 48 09" src="https://cloud.githubusercontent.com/assets/3933654/26718459/63e43a80-4781-11e7-8bd4-f7933e013054.png">
<img width="574" alt="capture d ecran 2017-06-02 a 10 48 00" src="https://cloud.githubusercontent.com/assets/3933654/26718458/63e2cf7e-4781-11e7-96e9-ea909d72f681.png">

:(

**Comment by loicjaouen** [comment]
- (empirically) building "elaborated toolbar"  doesn’t work when passed as config in `CKEDITOR.replace(tag, config);`
- definition of events can only be sent on `CKEDITOR.replace` config

b091ec31881c11f3f727c23aac1f808b63af8feb is splitting the two aparts.

<img width="839" alt="capture d ecran 2017-06-02 a 17 34 19" src="https://cloud.githubusercontent.com/assets/3933654/26733219/69f37074-47ba-11e7-8c91-6f6ca942d51b.png">

this leaves me with a question of javascript's variable scope about what happens if we want two ckEditors with different configs on the same page by redefining:
```
CKEDITOR.editorConfig = function( config ) {
```

... question that tears my brain apart, and the result should be checked empirically.  
Safety says that the config that differs should be passed to `CKEDITOR.replace`.

**Comment by loicjaouen** [comment]
Of course, at this stage, a complex html string can not be parsed by the knora's standoff parser.  
the next stage is the standoff mapping.

**Comment by benjamingeer** [comment]
@loicjaouen I don't know whether you've talked to Tobias about this, but I'm pretty sure he is using footnotes in the BEOL project and may have some suggestions about the standoff mapping.

**Comment by tobiasschweizer** [comment]
Currently, footnotes are **not** stored in a special way in the BEOL project. They are encoded using `<ref>` embedded in the text. This has the disadvantage that you cannot not search for whole phrases using double quotes when the footnote's text is in-between. However, I chose this approach because then the XML is very similar to the source code I am importing from (Wiki syntax) and therefore seems less error prone to me.

Here is the mapping I am using for BEOL: 
Mapping: <https://github.com/dhlab-basel/Knora/blob/develop/webapi/_test_data/test_route/texts/mappingForLetter.xml>

Simple Example:

```
This is my text<ref>footnote text</ref> and some other.
```

Before sending the XML to the client, it is converted to HTML using an XSL transformation.

<img width="1376" alt="footnote_in_html" src="https://cloud.githubusercontent.com/assets/6000023/26830312/bdd357fc-4ac8-11e7-8533-2b3db98560aa.png">

So texts are primarily stored in XML and converted to HTML when necessary (creating internal links). However, they would be edited in XML only (HTML is a derived format for read-only purpose).

If you want to use a footnote plugin to create HTML with footnotes that are actually stored at the end of the document, you would have to add the additional markup to the standoff mapping and the filter of CKeditor. I do not know how complicated that would be. It would be important that we could validate internal references. I think the `knora-base:StandoffInternalReferenceTag` (<http://www.knora.org/documentation/manual/rst/knora-api-server/api_v1/create-a-mapping.html#standoff-data-types>) should be used for this. 

If you feel that this mapping should be project specific, you would have to think about both project specific mappings and configs for CKeditor. Each `TextValue` is stored with a reference to the mapping used for conversion to and from standoff. But CKeditor is always called with the standard mapping and the GUI does not know about any other mappings.  

Maybe it would be best to look at the HTML produced by CKeditor's footnote plugin first.

**Comment by loicjaouen** [comment]
@tobiasschweizer : 

Thanks for the beol mapping.

We could translate footnotes into embedded `<ref>` tags as you did, even though it makes it hard to cite twice the same footnotes.

> If you want to use a footnote plugin

I will look at the internal links and at the html produced by ckeditor...

> If you feel that this mapping should be project specific

I don't necessarily want it project specific, it might be useful elsewhere.  
The problem might be to have a mix of ckeditor instances with different configs.  
And refactoring `jquery.htmleditor.js` to make it configurable might be a pain.

@tobiasschweizer also said: https://github.com/dhlab-basel/Knora/pull/527#issuecomment-306486326

> I think for the thing you have in mind, project specific configurations would be well suited. However, the old SALSAH GUI does not support that. So you would have to find a way to implement that.
>
> This would involve:
> 
> -    a specific mapping
> -    a specific configuration for CKeditor (toolbar -> controls the markup produced by the editor)
> -    a specific filter for CKeditor (markup validation)
> 
> You could assign each project a configuration. When creating a new text value, CKeditor should be loaded with the project's configuration. When editing an existing text, the configuration used before would have to be reloaded.
